src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -350,39 +348,101 @@ private void WriteAsPaxSharedInternal(Span<byte> buffer, out long actualLength)
+            {
+                if (_format is TarEntryFormat.V7)
+                {
+                    throw new ArgumentException(SR.Format(SR.TarEntryFieldExceedsMaxLength, nameof(TarEntry.Name)), ArgNameEntry);
+                }
+
+                int utf16NameTruncatedLength = GetUtf16TruncatedTextLength(name, FieldLengths.Name);
+                name = name.Slice(0, utf16NameTruncatedLength);
+            }
+
+            return WriteAsUtf8String(name, buffer.Slice(FieldLocations.Name, FieldLengths.Name));
         }
 
-        // Ustar and PAX save in the name byte array only the ASCII bytes that fit, and the rest of that string is saved in the prefix field.
-        private int WritePosixName(Span<byte> buffer)
+        // 'https://www.freebsd.org/cgi/man.cgi?tar(5)'
+        // If the pathname is too long to fit in the 100 bytes provided by the standard format,
+        // it can be split at any / character with the first portion going into the prefix field.
+        private int WriteUstarName(Span<byte> buffer)
         {
-            int checksum = WriteName(buffer);
+            // We can have a pathname as big as 256, prefix + '/' + name,
+            // the separator in between can be neglected as the reader will append it when it joins both fields.
+            const int MaxPathname = FieldLengths.Prefix + FieldLengths.Name + 1;
 
-            if (_name.Length > FieldLengths.Name)
+            if (GetUtf8TextLength(_name) > MaxPathname)
             {
-                int prefixBytesLength = Math.Min(_name.Length - FieldLengths.Name, FieldLengths.Prefix);
-                Span<byte> remaining = stackalloc byte[prefixBytesLength];
-                int encoded = Encoding.ASCII.GetBytes(_name.AsSpan(FieldLengths.Name, prefixBytesLength), remaining);
-                Debug.Assert(encoded == remaining.Length);
+                throw new ArgumentException(SR.TarEntryFieldExceedsMaxLength, ArgNameEntry);
+            }
+
+            Span<byte> encodingBuffer = stackalloc byte[MaxPathname];
+            int encoded = Encoding.UTF8.GetBytes(_name, encodingBuffer);
+            ReadOnlySpan<byte> pathnameBytes = encodingBuffer.Slice(0, encoded);
 
-                checksum += WriteLeftAlignedBytesAndGetChecksum(remaining, buffer.Slice(FieldLocations.Prefix, FieldLengths.Prefix));
+            // If the pathname is able to fit in Name, we can write it down there and avoid calculating Prefix.
+            if (pathnameBytes.Length <= FieldLengths.Name)
+            {
+                return WriteLeftAlignedBytesAndGetChecksum(pathnameBytes, buffer.Slice(FieldLocations.Name, FieldLengths.Name));
             }
 
-            return checksum;
+            int lastIdx = pathnameBytes.LastIndexOfAny(PathInternal.Utf8DirectorySeparators);
+            scoped ReadOnlySpan<byte> name;
+            scoped ReadOnlySpan<byte> prefix;
+
+            if (lastIdx < 1) // splitting at the root is not allowed.
```

Comment:
> // splitting at the root is not allowed.

I copied this behavior from GNU Tar, it does not store one single separtor in the prefix, I assume we should do the same. 

No big deal to go the other way but "absolute path" is already very odd scenario for tar and I don't have a reason to not do the same. 

On top of that, I think we should instruct users to be aware about/avoid using absolute paths with Tar APIs.